state/unit.go

```diff
@@ -321,6 +321,103 @@ func (u *Unit) PasswordValid(password string) bool {
 	return false
 }
 
+// unitStateDoc records the state persisted by the charm executing in the unit.
+type unitStateDoc struct {
+	// DocID is always the same as a unit's global key.
+	DocID string `bson:"_id"`
+
+	// State encodes the unit's persisted state as a list of key-value pairs.
+	State map[string]string `bson:"state,omitempty"`
+
+	TxnRevno int64 `bson:"txn-revno"`
+}
+
+// removeUnitStateOp returns the operation needed to remove the unit state
+// document associated with the given globalKey.
+func removeUnitStateOp(mb modelBackend, globalKey string) txn.Op {
+	return txn.Op{
+		C:      unitStatesC,
+		Id:     mb.docID(globalKey),
+		Remove: true,
+	}
+}
+
+// SetState persisted the state for a unit.
+func (u *Unit) SetState(unitState map[string]string) error {
+	unitGlobalKey := u.globalKey()
+	txnDoc := struct {
+		TxnRevno int64 `bson:"txn-revno"`
+	}{}
+
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		coll, closer := u.st.db().GetCollection(unitStatesC)
+		defer closer()
+
```

Comment:
We should ensure the unit is alive as part of the transaction in general.

Reply:
That's a valid point. I will add an assertion and a test for this case